File: php/tests/KiahkTest.php
<?php

declare(strict_types=1);

namespace Wizardlabz\Kiahk\Tests;

use PHPUnit\Framework\Attributes\DataProvider;
use PHPUnit\Framework\TestCase;
use Wizardlabz\Kiahk\Algorithms;
use Wizardlabz\Kiahk\CopticCalendar;
use Wizardlabz\Kiahk\CopticDate;
use Wizardlabz\Kiahk\Exception\InvalidCopticDateException;
use Wizardlabz\Kiahk\Exception\InvalidCopticMonthException;
use Wizardlabz\Kiahk\Exception\InvalidGregorianDateException;
use Wizardlabz\Kiahk\Exception\UnknownFeastException;
use Wizardlabz\Kiahk\Exception\UnsupportedLocaleException;
use Wizardlabz\Kiahk\GregorianDate;

final class KiahkTest extends TestCase
{
    /** @return array<string,mixed> */
    private static function vectors(): array
    {
        // Load the shared cross-port test contract from core/test-vectors.json.
        // __DIR__ is php/tests/ → ../../core/test-vectors.json (repo root).
        $path = __DIR__ . '/../../core/test-vectors.json';
        $json = file_get_contents($path);
        if ($json === false) {
            throw new \RuntimeException("Failed to read $path");
        }
        return json_decode($json, true, flags: JSON_THROW_ON_ERROR);
    }

    // -------------------------------------------------------------------------
    // gregorian ↔ coptic (data-driven from core/test-vectors.json)
    // -------------------------------------------------------------------------

    public static function gregorianToCopticProvider(): iterable
    {
        foreach (self::vectors()['gregorian_to_coptic'] as $i => $v) {
            yield "vector#$i" => [$v['gregorian'], $v['coptic']];
        }
    }

    #[DataProvider('gregorianToCopticProvider')]
    public function testGregorianToCoptic(array $g, array $expected): void
    {
        $c = (new GregorianDate($g['year'], $g['month'], $g['day']))->toCoptic();
        self::assertSame($expected['year'], $c->year);
        self::assertSame($expected['month'], $c->month);
        self::assertSame($expected['day'], $c->day);
    }

    public static function copticToGregorianProvider(): iterable
    {
        foreach (self::vectors()['coptic_to_gregorian'] as $i => $v) {
            yield "vector#$i" => [$v['coptic'], $v['gregorian']];
        }
    }

    #[DataProvider('copticToGregorianProvider')]
    public function testCopticToGregorian(array $c, array $expected): void
    {
        $g = (new CopticDate($c['year'], $c['month'], $c['day']))->toGregorian();
        self::assertSame($expected['year'], $g->year);
        self::assertSame($expected['month'], $g->month);
        self::assertSame($expected['day'], $g->day);
    }

    // -------------------------------------------------------------------------
    // Round-trip identity over a long range
    // -------------------------------------------------------------------------

    public function testRoundTripGregorianCopticGregorian(): void
    {
        // Sample dates spanning 1900..2099 — every conversion must round-trip.
        $cases = [
            [1900, 1, 1], [1950, 6, 15], [2000, 2, 29], [2024, 12, 31],
            [2025, 1, 11], [2050, 7, 4], [2099, 12, 31],
        ];
        foreach ($cases as [$y, $m, $d]) {
            $g = new GregorianDate($y, $m, $d);
            $roundTrip = $g->toCoptic()->toGregorian();
            self::assertTrue($g->equals($roundTrip), "round-trip failed for $y-$m-$d");
        }
    }

    // -------------------------------------------------------------------------
    // Easter
    // -------------------------------------------------------------------------

    public static function easterProvider(): iterable
    {
        foreach (self::vectors()['easter'] as $v) {
            yield "easter_{$v['gregorian_year']}" => [$v['gregorian_year'], $v['date']];
        }
    }

    #[DataProvider('easterProvider')]
    public function testEaster(int $gYear, array $expected): void
    {
        $easter = CopticCalendar::easterDate($gYear);
        self::assertSame($expected['year'], $easter->year);
        self::assertSame($expected['month'], $easter->month);
        self::assertSame($expected['day'], $easter->day);
    }

    // -------------------------------------------------------------------------
    // Moveable feasts
    // -------------------------------------------------------------------------

    public static function moveableFeastProvider(): iterable
    {
        foreach (self::vectors()['moveable_feasts'] as $i => $v) {
            yield "moveable#$i" => [$v['feast_id'], $v['gregorian_year'], $v['date']];
        }
    }

    #[DataProvider('moveableFeastProvider')]
    public function testMoveableFeast(string $id, int $gYear, array $expected): void
    {
        $feast = CopticCalendar::moveableFeast($id, $gYear);
        $g = $feast->gregorianDate;
        self::assertSame($expected['year'], $g->year);
        self::assertSame($expected['month'], $g->month);
        self::assertSame($expected['day'], $g->day);
        self::assertSame($id, $feast->id());
        self::assertSame('moveable', $feast->type());
    }

    public function testUnknownMoveableFeastThrows(): void
    {
        $this->expectException(UnknownFeastException::class);
        CopticCalendar::moveableFeast('not_a_real_feast', 2025);
    }

    // -------------------------------------------------------------------------
    // Year feasts (full list, sorted)
    // -------------------------------------------------------------------------

    public function testYearFeasts2025IsSortedAndIncludesAllMajor(): void
    {
        $feasts = CopticCalendar::yearFeasts(2025);
        // 6 fixed + 6 moveable (great_lent is in core/feasts.json) — but the
        // exact count depends on which fixed feasts actually land in 2025.
        self::assertGreaterThanOrEqual(11, count($feasts));

        $prev = null;
        foreach ($feasts as $feast) {
            $g = $feast->gregorianDate;
            $jdn = Algorithms::gregorianToJdn($g->year, $g->month, $g->day);
            if ($prev !== null) {
                self::assertGreaterThanOrEqual($prev, $jdn, 'feasts must be sorted ascending');
            }
            $prev = $jdn;
        }

        // Sanity: Easter Sunday must be in the list for 2025
        $ids = array_map(fn($f) => $f->id(), $feasts);
        self::assertContains('easter', $ids);
        self::assertContains('nativity', $ids);
    }

    // -------------------------------------------------------------------------
    // Coptic month names (data-driven where the contract provides them)
    // -------------------------------------------------------------------------

    public static function copticMonthNamesProvider(): iterable
    {
        foreach (self::vectors()['coptic_month_names'] ?? [] as $i => $v) {
            yield "month#$i" => [$v['month'], $v['locale'], $v['name']];
        }
    }

    #[DataProvider('copticMonthNamesProvider')]
    public function testMonthName(int $month, string $locale, string $expected): void
    {
        self::assertSame($expected, CopticCalendar::monthName($month, $locale));
    }

    public function testInvalidMonthThrows(): void
    {
        $this->expectException(InvalidCopticMonthException::class);
        CopticCalendar::monthName(14, 'en');
    }

    public function testUnsupportedLocaleThrows(): void
    {
        $this->expectException(UnsupportedLocaleException::class);
        CopticCalendar::monthName(1, 'fr');
    }

    // -------------------------------------------------------------------------
    // Validation
    // -------------------------------------------------------------------------

    public static function invalidGregorianProvider(): iterable
    {
        foreach (self::vectors()['invalid_gregorian_dates'] as $i => $v) {
            yield "bad_g#$i" => [$v['year'], $v['month'], $v['day']];
        }
    }

    #[DataProvider('invalidGregorianProvider')]
    public function testInvalidGregorianDatesAreRejected(int $y, int $m, int $d): void
    {
        $this->expectException(InvalidGregorianDateException::class);
        new GregorianDate($y, $m, $d);
    }

    public static function invalidCopticProvider(): iterable
    {
        foreach (self::vectors()['invalid_coptic_dates'] as $i => $v) {
            yield "bad_c#$i" => [$v['year'], $v['month'], $v['day']];
        }
    }

    #[DataProvider('invalidCopticProvider')]
    public function testInvalidCopticDatesAreRejected(int $y, int $m, int $d): void
    {
        $this->expectException(InvalidCopticDateException::class);
        new CopticDate($y, $m, $d);
    }

    // -------------------------------------------------------------------------
    // DateTime interop
    // -------------------------------------------------------------------------

    public function testDateTimeImmutableInterop(): void
    {
        $g = new GregorianDate(2025, 1, 11);
        $dt = $g->toDateTimeImmutable();
        self::assertSame('2025-01-11', $dt->format('Y-m-d'));
        self::assertTrue($g->equals(GregorianDate::fromDateTimeImmutable($dt)));
    }
}

File: kotlin/src/test/kotlin/com/wizardlabz/kiahk/KiahkTest.kt
package com.wizardlabz.kiahk

import com.google.gson.JsonArray
import com.google.gson.JsonObject
import com.google.gson.JsonParser
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.DynamicTest
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.TestFactory
import java.io.File
import java.time.LocalDate

/**
 * Cross-port test suite — every assertion is data-driven from
 * `core/test-vectors.json`. Identical contract to the JS / Python / Go /
 * Dart / Swift / C# / C / PHP ports.
 */
class KiahkTest {

    // -------------------------------------------------------------------------
    // Test vector loading
    // -------------------------------------------------------------------------

    private companion object {
        // Gradle runs tests with cwd = the `kotlin/` module directory,
        // so the shared spec is one level up at ../core/test-vectors.json.
        val vectors: JsonObject by lazy {
            val text = File("../core/test-vectors.json").readText()
            JsonParser.parseString(text).asJsonObject
        }
    }

    // -------------------------------------------------------------------------
    // Gregorian ↔ Coptic
    // -------------------------------------------------------------------------

    @TestFactory
    fun gregorianToCoptic(): List<DynamicTest> =
        vectors.array("gregorian_to_coptic").mapIndexed { i, raw ->
            DynamicTest.dynamicTest("vector#$i") {
                val v = raw.asJsonObject
                val g = v.obj("gregorian"); val exp = v.obj("coptic")
                val c = GregorianDate(g.int("year"), g.int("month"), g.int("day")).toCoptic()
                assertEquals(exp.int("year"), c.year)
                assertEquals(exp.int("month"), c.month)
                assertEquals(exp.int("day"), c.day)
            }
        }

    @TestFactory
    fun copticToGregorian(): List<DynamicTest> =
        vectors.array("coptic_to_gregorian").mapIndexed { i, raw ->
            DynamicTest.dynamicTest("vector#$i") {
                val v = raw.asJsonObject
                val c = v.obj("coptic"); val exp = v.obj("gregorian")
                val g = CopticDate(c.int("year"), c.int("month"), c.int("day")).toGregorian()
                assertEquals(exp.int("year"), g.year)
                assertEquals(exp.int("month"), g.month)
                assertEquals(exp.int("day"), g.day)
            }
        }

    // -------------------------------------------------------------------------
    // Round-trip identity
    // -------------------------------------------------------------------------

    @TestFactory
    fun roundTripGregorianCopticGregorian(): List<DynamicTest> {
        val cases = listOf(
            Triple(1900, 1, 1), Triple(1950, 6, 15), Triple(2000, 2, 29),
            Triple(2024, 12, 31), Triple(2025, 1, 11), Triple(2050, 7, 4),
            Triple(2099, 12, 31),
        )
        return cases.map { (y, m, d) ->
            DynamicTest.dynamicTest("$y-$m-$d") {
                val g = GregorianDate(y, m, d)
                assertEquals(g, g.toCoptic().toGregorian())
            }
        }
    }

    // -------------------------------------------------------------------------
    // Easter
    // -------------------------------------------------------------------------

    @TestFactory
    fun easter(): List<DynamicTest> =
        vectors.array("easter").map { raw ->
            val v = raw.asJsonObject
            val year = v.int("gregorian_year")
            DynamicTest.dynamicTest("easter_$year") {
                val exp = v.obj("date")
                val e = CopticCalendar.easterDate(year)
                assertEquals(exp.int("year"), e.year)
                assertEquals(exp.int("month"), e.month)
                assertEquals(exp.int("day"), e.day)
            }
        }

    // -------------------------------------------------------------------------
    // Moveable feasts
    // -------------------------------------------------------------------------

    @TestFactory
    fun moveableFeasts(): List<DynamicTest> =
        vectors.array("moveable_feasts").mapIndexed { i, raw ->
            DynamicTest.dynamicTest("moveable#$i") {
                val v = raw.asJsonObject
                val exp = v.obj("date")
                val feast = CopticCalendar.moveableFeast(
                    v["feast_id"].asString,
                    v.int("gregorian_year"),
                )
                assertEquals(exp.int("year"), feast.gregorianDate.year)
                assertEquals(exp.int("month"), feast.gregorianDate.month)
                assertEquals(exp.int("day"), feast.gregorianDate.day)
                assertEquals(v["feast_id"].asString, feast.id)
                assertEquals("moveable", feast.type)
            }
        }

    // -------------------------------------------------------------------------
    // Year feasts (sorting + completeness)
    // -------------------------------------------------------------------------

    @Test
    fun yearFeasts2025IsSortedAndIncludesMajor() {
        val feasts = CopticCalendar.yearFeasts(2025)
        assertTrue(feasts.size >= 11, "expected >= 11 major feasts in 2025, got ${feasts.size}")

        var prevJdn = Int.MIN_VALUE
        for (f in feasts) {
            val jdn = Algorithms.gregorianToJdn(f.gregorianDate.year, f.gregorianDate.month, f.gregorianDate.day)
            assertTrue(jdn >= prevJdn, "feasts must be sorted ascending")
            prevJdn = jdn
        }

        val ids = feasts.map { it.id }
        assertTrue("easter" in ids)
        assertTrue("nativity" in ids)
    }

    // -------------------------------------------------------------------------
    // Coptic month names (data-driven if the vector file exposes them)
    // -------------------------------------------------------------------------

    @TestFactory
    fun monthNames(): List<DynamicTest> {
        val list = vectors.getAsJsonArray("coptic_month_names") ?: return emptyList()
        return list.mapIndexed { i, raw ->
            val v = raw.asJsonObject
            DynamicTest.dynamicTest("month#$i") {
                assertEquals(
                    v["name"].asString,
                    CopticCalendar.monthName(v.int("month"), v["locale"].asString),
                )
            }
        }
    }

    @Test
    fun invalidMonthThrows() {
        assertThrows(InvalidCopticMonthException::class.java) {
            CopticCalendar.monthName(14, "en")
        }
    }

    @Test
    fun unsupportedLocaleThrows() {
        assertThrows(UnsupportedLocaleException::class.java) {
            CopticCalendar.monthName(1, "fr")
        }
    }

    @Test
    fun unknownMoveableFeastThrows() {
        assertThrows(UnknownFeastException::class.java) {
            CopticCalendar.moveableFeast("not_a_real_feast", 2025)
        }
    }

    // -------------------------------------------------------------------------
    // Validation negatives
    // -------------------------------------------------------------------------

    @TestFactory
    fun invalidGregorianDatesAreRejected(): List<DynamicTest> =
        vectors.array("invalid_gregorian_dates").mapIndexed { i, raw ->
            DynamicTest.dynamicTest("bad_g#$i") {
                val v = raw.asJsonObject
                assertThrows(InvalidGregorianDateException::class.java) {
                    GregorianDate(v.int("year"), v.int("month"), v.int("day"))
                }
            }
        }

    @TestFactory
    fun invalidCopticDatesAreRejected(): List<DynamicTest> =
        vectors.array("invalid_coptic_dates").mapIndexed { i, raw ->
            DynamicTest.dynamicTest("bad_c#$i") {
                val v = raw.asJsonObject
                assertThrows(InvalidCopticDateException::class.java) {
                    CopticDate(v.int("year"), v.int("month"), v.int("day"))
                }
            }
        }

    // -------------------------------------------------------------------------
    // LocalDate interop
    // -------------------------------------------------------------------------

    @Test
    fun localDateInterop() {
        val g = GregorianDate(2025, 1, 11)
        val ld = g.toLocalDate()
        assertEquals(LocalDate.of(2025, 1, 11), ld)
        assertEquals(g, GregorianDate.fromLocalDate(ld))
    }
}

private fun JsonObject.array(key: String): JsonArray = getAsJsonArray(key)
private fun JsonObject.obj(key: String): JsonObject = getAsJsonObject(key)
private fun JsonObject.int(key: String): Int = get(key).asInt
